core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -1367,9 +1368,108 @@ public AlluxioURI translateUri(String uriStr) throws InvalidPathException {
     // Scheme-less URIs are regarded as Alluxio URI.
     if (uri.getScheme() == null || uri.getScheme().equals(Constants.SCHEME)) {
       return uri;
-    } else {
-      // Reverse lookup mount table to find mount point that contains the path.
+    }
+
+    // Reverse lookup mount table to find mount point that contains the path.
+    try {
       return mMountTable.translate(uriStr);
+    } catch (InvalidPathException e) {
+      // Throw if auto-mount is not requested.
+      if (!ServerConfiguration.getBoolean(PropertyKey.MASTER_AUTO_MOUNT_ENABLED)) {
+        throw e;
+      }
+    }
+
+    // Reverse lookup failed and auto-mount is requested.
+    try {
+      autoMountUfsForUri(uri);
+    } catch (Exception e) {
+      LOG.warn(String.format("Failed to auto-mount ufs for URI: %s", uri), e);
+      if (e instanceof InvalidPathException) {
+        // Don't throw as this could be thrown due to ufs having been mounted concurrently.
+      } else {
+        throw new InvalidPathException(
+            String.format("Failed to translate and auto-mount path: %s", uriStr), e);
+      }
+    }
+    // Try translating the URI after auto-mount.
+    return mMountTable.translate(uriStr);
+  }
+
+  /**
+   * Attempts to mount given UFS Uri into designated auto-mount directory.
+   *
+   * Mount attempt will start from the mounting lowest path component, and go to higher path
+   * components if mounting fails due to access control violation.
+   *
+   * TODO(ggezer) Have fine-grained locking on mount paths instead of marking synchronized.
+   *
+   * @param ufsUri ufs Uri for auto-mount
+   * @throws AlluxioException
+   * @throws IOException
+   */
+  private synchronized void autoMountUfsForUri(AlluxioURI ufsUri)
+      throws AlluxioException, IOException {
+    // Generate lowest Alluxio mount root for given path.
+    String alluxioMountRoot =
+        PathUtils.concatPath(ServerConfiguration.get(PropertyKey.MASTER_AUTO_MOUNT_ROOT),
+            ufsUri.getScheme(), ufsUri.getAuthority());
+    // Read default options for auto-mounted UFSes.
+    boolean mountReadonly = ServerConfiguration.getBoolean(PropertyKey.MASTER_AUTO_MOUNT_READONLY);
+    boolean mountShared = ServerConfiguration.getBoolean(PropertyKey.MASTER_AUTO_MOUNT_SHARED);
+    // Try mounting UFS to Alluxio starting from its lowest path.
+    int pathComponentIndex = 0;
+    String currentPathComponent;
+    // Used to keep track of first directory that is created for this auto-mount call.
+    // It'll be deleted recursively if this mount fails.
+    AlluxioURI firstCreatedParent = null;
+    try {
+      while ((currentPathComponent = ufsUri.getLeadingPath(pathComponentIndex++)) != null) {
```

Comment:
can we add some comments on the algorithm? Like if we access "ufs://a/b/c/d" and "ufs://a/b/c" is the first accessible one, which ufs path is mounted to which alluxio path?